homeassistant/components/braviatv/coordinator.py

```diff
@@ -253,6 +253,15 @@ async def async_select_source(self, source: str) -> None:
     @catch_braviatv_errors
     async def async_send_command(self, command: Iterable[str], repeats: int) -> None:
         """Send command to device."""
+        commands = await self.client.get_command_list()
         for _ in range(repeats):
             for cmd in command:
-                await self.client.send_command(cmd)
+                if cmd in commands:
+                    await self.client.send_command(cmd)
+                else:
+                    available_commands = ", ".join(commands.keys())
+                    _LOGGER.error(
```

Comment:
I recommend raising a ValueError with the message as exception argument instead. The message will then show in the frontend. We prefer doing this nowadays over logging errors in service handlers.

Reply:
For some reason the message from the ValueError exception thrown in remote.send_command is showing as Unknown error in frontend. At the same time, 2 errors appear in the log, in which it is difficult to find our message.

```py
async def async_send_command(self, command: Iterable[str], repeats: int) -> None:
    """Send command to device."""
    raise ValueError("Test")
```

logs:

```
[4695602416] Error handling message: Unknown error (unknown_error)
Traceback (most recent call last):
  File "/Users/draft/Test/home-assistant/homeassistant/components/websocket_api/decorators.py", line 27, in _handle_async_response
    await func(hass, connection, msg)
  File "/Users/draft/Test/home-assistant/homeassistant/components/websocket_api/commands.py", line 647, in handle_execute_script
    await script_obj.async_run(msg.get("variables"), context=context)
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 1513, in async_run
    await asyncio.shield(run.async_run())
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 405, in async_run
    await self._async_step(log_exceptions=False)
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 449, in _async_step
    self._handle_exception(
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 472, in _handle_exception
    raise exception
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 447, in _async_step
    await getattr(self, handler)()
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 680, in _async_call_service_step
    await service_task
  File "/Users/draft/Test/home-assistant/homeassistant/core.py", line 1738, in async_call
    task.result()
  File "/Users/draft/Test/home-assistant/homeassistant/core.py", line 1775, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await service.entity_service_call(
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/service.py", line 676, in entity_service_call
    future.result()  # pop exception if have
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/entity.py", line 931, in async_request_call
    await coro
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/service.py", line 713, in _handle_entity_call
    await result
  File "/Users/draft/Test/home-assistant/homeassistant/components/braviatv/remote.py", line 49, in async_send_command
    await self.coordinator.async_send_command(command, repeats)
  File "/Users/draft/Test/home-assistant/homeassistant/components/braviatv/coordinator.py", line 43, in wrapper
    await func(self, *args, **kwargs)
  File "/Users/draft/Test/home-assistant/homeassistant/components/braviatv/coordinator.py", line 256, in async_send_command
    raise ValueError("Test")
ValueError: Test
```

```
websocket_api script: Error executing script. Unexpected error for call_service at pos 1: Test
Traceback (most recent call last):
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 447, in _async_step
    await getattr(self, handler)()
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/script.py", line 680, in _async_call_service_step
    await service_task
  File "/Users/draft/Test/home-assistant/homeassistant/core.py", line 1738, in async_call
    task.result()
  File "/Users/draft/Test/home-assistant/homeassistant/core.py", line 1775, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await service.entity_service_call(
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/service.py", line 676, in entity_service_call
    future.result()  # pop exception if have
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/entity.py", line 931, in async_request_call
    await coro
  File "/Users/draft/Test/home-assistant/homeassistant/helpers/service.py", line 713, in _handle_entity_call
    await result
  File "/Users/draft/Test/home-assistant/homeassistant/components/braviatv/remote.py", line 49, in async_send_command
    await self.coordinator.async_send_command(command, repeats)
  File "/Users/draft/Test/home-assistant/homeassistant/components/braviatv/coordinator.py", line 43, in wrapper
    await func(self, *args, **kwargs)
  File "/Users/draft/Test/home-assistant/homeassistant/components/braviatv/coordinator.py", line 256, in async_send_command
    raise ValueError("Test")
ValueError: Test
```
Throwing an exception for example from select_source is displayed correctly. Can you tell me what is the problem?

Reply:
I made changes but the problem remains.

Reply:
Doesn't the message, ie exception argument, show in the frontend after it says unknown error?

Reply:
> Doesn't the message, ie exception argument, show in the frontend after it says unknown error?

Nope, only unknown error. 

https://user-images.githubusercontent.com/11841379/187647578-6a9c0498-a712-4ea1-9592-42dd057f7d70.mov

Reply:
I've asked the frontend team if they have a clue. This should work.

Reply:
If you print the exception instance before raising it, does it print as expected?

Reply:
```py
exc = ValueError(
    f"Unsupported command: {cmd}, list of available commands: {commands_keys}"
)
print(exc)
```

Yes, the message prints correctly in logs:
```
Unsupported command: Test, list of available commands: Num1, Num2, Num3, Num4, Num5, Num6, Num7, Num8, Num9, Num0, Num11, Num12, Enter, GGuide, ChannelUp, ChannelDown, VolumeUp, VolumeDown, Mute, TvPower, Audio, MediaAudioTrack, Tv, Input, TvInput, TvAntennaCable, WakeUp, PowerOff, Sleep, Right, Left, SleepTimer, Analog2, TvAnalog, Display, Jump, PicOff, PictureOff, Teletext, Video1, Video2, AnalogRgb1, Home, Exit, PictureMode, AdvancedBSCS, Confirm, Up, Down, ClosedCaption, Component1, Component2, Wide, EPG, PAP, TenKey, BSCS, Ddata, Stop, Pause, Play, Rewind, Forward, DOT, Rec, Return, Blue, Red, Green, Yellow, SubTitle, CS, BS, Digital, Options, Media, Prev, Next, DpadCenter, CursorUp, CursorDown, CursorLeft, CursorRight, ShopRemoteControlForcedDynamic, FlashPlus, FlashMinus, DemoMode, Analog, Mode3D, DigitalToggle, DemoSurround, *AD, AudioMixUp, AudioMixDown, PhotoFrame, Tv_Radio, SyncMenu, Hdmi1, Hdmi2, Hdmi3, Hdmi4, TopMenu, PopUpMenu, OneTouchTimeRec, OneTouchView, FootballMode, iManual, Netflix, Assists, FeaturedApp, FeaturedAppVOD, GooglePlay, ActionMenu, Help, TvSatellite, WirelessSubwoofer, AndroidMenu, RecorderMenu, STBMenu, MuteOn, MuteOff, AudioOutput_AudioSystem, AudioOutput_TVSpeaker, AudioOutput_Toggle, ApplicationLauncher, YouTube, PartnerApp1, PartnerApp2, PartnerApp3, PartnerApp4, PartnerApp5, PartnerApp6, PartnerApp7, PartnerApp8, PartnerApp9, PartnerApp10, PartnerApp11, PartnerApp12, PartnerApp13, PartnerApp14, PartnerApp15, PartnerApp16, PartnerApp17, PartnerApp18, PartnerApp19, PartnerApp20
```

Reply:
This is currently a consequence of the dev tools implementation in the frontend. The message will show in the normal frontend outside dev tools. Is there an entity card for remote entities?

Reply:
As I know, there is no special remote entity card for sending commands.

Reply:
If wrap the call in a script and run it, the message is displayed.

<img width="600" src="https://user-images.githubusercontent.com/11841379/187662518-ee799192-0655-4d23-a47a-190413b91bc2.png">

Reply:
P.S. It's hard to read such a long text as the popup disappears after 3 seconds even though I hover over it.

Reply:
It's ok to log an error if that's more user friendly, for now. :+1: